cmd/discovery/client/client.go

```diff
@@ -63,114 +71,234 @@ func (c *Client) Start() error {
+	attestationEndpoint := fmt.Sprintf("%s/tpm-attestation", serverURL)
+	c.Logger.Debugf("Debug: TPM attestation endpoint: %s", attestationEndpoint)
+
+	// Step 1: Initialize Remote Attestation Client (outside the retry loop)
+	c.Logger.Debugf("Debug: Initializing Remote Attestation Client")
+	clientOpts := []tpm.Option{}
+	if c.Config.Kcrypt.Challenger.TPMDevice != "" {
+		c.Logger.Debugf("Debug: Using TPM device: %s", c.Config.Kcrypt.Challenger.TPMDevice)
+		clientOpts = append(clientOpts, tpm.WithTPMDevice(c.Config.Kcrypt.Challenger.TPMDevice))
+	}
+	attestationClient, err := attestation.NewRemoteAttestationClient(clientOpts...)
 	if err != nil {
-		return err
+		return "", fmt.Errorf("failed to create attestation client: %w", err)
 	}
-	bpass := base64.RawURLEncoding.EncodeToString(pass)
+	c.Logger.Debugf("Debug: Remote Attestation Client initialized successfully")
 
-	opts := []tpm.Option{
-		tpm.WithCAs([]byte(c.Config.Kcrypt.Challenger.Certificate)),
-		tpm.AppendCustomCAToSystemCA,
-		tpm.WithAdditionalHeader("label", p.FilesystemLabel),
-		tpm.WithAdditionalHeader("name", p.Name),
-		tpm.WithAdditionalHeader("uuid", p.UUID),
+	// Ensure client is properly closed when done
+	defer func() {
+		if closeErr := attestationClient.Close(); closeErr != nil {
+			c.Logger.Debugf("Warning: Failed to close attestation client: %v", closeErr)
+		}
+	}()
+
+	var lastErr error
+	for tries := range attempts {
+		c.Logger.Debugf("Debug: TPM attestation attempt %d/%d", tries+1, attempts)
+
+		// Step 2: Start WebSocket-based attestation flow
+		c.Logger.Debugf("Debug: Starting WebSocket-based attestation flow")
+		passphrase, err := c.performTPMAttestation(attestationEndpoint, additionalHeaders, attestationClient, p)
+		if err != nil {
+			c.Logger.Debugf("Failed TPM attestation: %v", err)
+			lastErr = err
+
+			// Don't retry on attestation failures (security rejections like quarantine, PCR mismatch, etc.)
+			// These are permanent failures, not transient network issues
+			if strings.Contains(err.Error(), "attestation failed") {
+				c.Logger.Debugf("Attestation failure detected, not retrying")
+				return "", err
+			}
+
+			time.Sleep(NetworkRetryDelay)
+			continue
+		}
+
+		return passphrase, nil
 	}
-	for k, v := range headers {
-		opts = append(opts, tpm.WithAdditionalHeader(k, v))
+
+	if lastErr != nil {
+		return "", fmt.Errorf("exhausted all attempts (%d) for TPM attestation, last error: %w", attempts, lastErr)
 	}
+	return "", fmt.Errorf("exhausted all attempts (%d) for TPM attestation", attempts)
+}
 
-	conn, err := tpm.Connection(postEndpoint, opts...)
+// isLiveCDMode checks if the system is running in livecd mode
+// using kairos-sdk state detection (same as `kairos-agent state get boot`)
+func isLiveCDMode(logger types.KairosLogger) bool {
+	// Create a zerolog logger from the kairos logger for compatibility
+	runtime, err := state.NewRuntime()
```

Comment:
you should do `NewRuntimeWithLogger(logger)` so it uses the same logger instead of instantiating a new one

Reply:
True! Fixing.